jupyterhubutils/lsstmgr/namespacemanager.py

```diff
@@ -0,0 +1,298 @@
+    '''
+    namespace = None
+    service_account = None
+
+    def __init__(self, *args, **kwargs):
+        self.log = make_logger()
+        self.log.debug("Creating LSSTNamespaceManager")
+        self.parent = kwargs.pop('parent')
+
+    def update_namespace_name(self):
+        '''Build namespace name from user and execution namespace.
+        '''
+        execution_namespace = get_execution_namespace()
+        self.log.debug("Execution namespace: '{}'".format(execution_namespace))
+        user = self.parent.user
+        self.log.debug("User: '{}'".format(user))
+        username = user.escaped_name
+        if execution_namespace and username:
+            self.namespace = "{}-{}".format(execution_namespace,
+                                            username)
+        else:
+            df_msg = "Using 'default' namespace."
+            self.log.warning(df_msg)
+            self.namespace = "default"
+
+    def ensure_namespace(self):
+        '''Here we make sure that the namespace exists, creating it if
+        it does not.  That requires a ClusterRole that can list and create
+        namespaces.
+
+        If we have shadow PVs, we clone the (static) NFS PVs and then
+        attach namespaced PVCs to them.  Thus the role needs to be
+        able to list and create PVs and PVCs.
+
+        If we create the namespace, we also create (if needed) a ServiceAccount
+        within it to allow the user pod to spawn dask and workflow pods.
+
+        '''
+        self.update_namespace_name()
+        namespace = self.namespace
+        api = self.parent.api
+        if namespace == "default":
+            self.log.warning("Namespace is 'default'; no manipulation.")
+            return
+        ns = client.V1Namespace(
+            metadata=client.V1ObjectMeta(name=namespace))
+        try:
+            self.log.info("Attempting to create namespace '%s'" % namespace)
+            api.create_namespace(ns)
+        except ApiException as e:
+            if e.status != 409:
+                estr = "Create namespace '%s' failed: %s" % (ns, str(e))
+                self.log.exception(estr)
+                raise
+            else:
+                self.log.info("Namespace '%s' already exists." % namespace)
+        # Wait for the namespace to actually appear before creating objects
+        #  in it.
+        self._wait_for_namespace()
+        if self.service_account:
+            self.log.debug("Ensuring namespaced service account.")
+            self._ensure_namespaced_service_account()
+        else:
+            self.log.debug("No namespaced service account required.")
+        if self.parent.spawner.enable_namespace_quotas:
+            self.log.debug("Determining resource quota.")
+            qm = self.parent.quota_mgr
+            quota = qm.get_resource_quota_spec()
+            if quota:
+                self.log.debug("Ensuring namespace quota.")
+                qm.ensure_namespaced_resource_quota(quota)
```

Comment:
could probably condense this block if ensure_namespaced_resource_quota just looked at get_resource_quota_spec itself and checked if it should exist